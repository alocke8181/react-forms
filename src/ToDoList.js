import React,{useState} from "react";
import {v4 as uuid} from 'uuid';

import Todo from './Todo';
import NewToDoForm from "./NewToDoForm";

const ToDoList = ()=>{
    let [todos, setTodos] = useState([])

    const addTodo = (text) =>{
        setTodos(todos => [...todos, {key:uuid(), text}])
    }

    const removeTodo = (todoKey) =>{
        setTodos(todos => todos.filter((todo) => todo.key != todoKey))
    }

    return(
        <div>
            <NewToDoForm addTodo={addTodo} />
            <div>
                <ul>
                    {todos.map((todo) => <Todo key={todo.key} id={todo.key} text={todo.text}  removeTodo={removeTodo}/>)}
                </ul>
            </div>
        </div>
    )

}

export default ToDoList