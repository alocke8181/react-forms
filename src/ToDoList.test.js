import { render, fireEvent, screen, getByText } from '@testing-library/react';
import ToDoList from './ToDoList';

test('smoke',()=>{
    render(<ToDoList/>)
})

test('snapshot',()=>{
    const {asFragment} = render(<ToDoList/>)
    expect(asFragment()).toMatchSnapshot();
})

test('adding/removing todos', ()=>{
    const {container} = render(<ToDoList/>)
    const submitButton = container.querySelector('button[id="todo-submit"]');

    //expect the text field to be blank
    let todoBox = screen.getByLabelText("New Todo Item:").closest('input')
    expect(todoBox.value).toEqual('');

    //fill the form data
    fireEvent.change(todoBox, {target:{value:'test'}});

    //expect the form value to change
    todoBox = screen.getByLabelText("New Todo Item:").closest('input')
    expect(todoBox.value).toEqual('test');

    //submit
    fireEvent.click(submitButton);

    //check for the new item
    expect(container.querySelector('li[class="todo-item"]')).toBeInTheDocument();
    expect(getByText(container, 'test')).toBeInTheDocument();

    //check the form has reset
    todoBox = screen.getByLabelText("New Todo Item:").closest('input')
    expect(todoBox.value).toEqual('');

    //click the remove button
    const removeButton = container.querySelector('button[class="todo-remove-button"]');
    fireEvent.click(removeButton);

    //check the todo is gone
    expect(container.querySelector('li[class="todo-item"]')).not.toBeInTheDocument();
})