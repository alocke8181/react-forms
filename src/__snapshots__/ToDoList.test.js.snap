// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<DocumentFragment>
  <div>
    <form>
      <label
        for="text"
      >
        New Todo Item:
      </label>
      <input
        id="text"
        name="text"
        placeholder=""
        type="text"
        value=""
      />
      <button
        id="todo-submit"
      >
        Submit
      </button>
    </form>
    <div>
      <ul />
    </div>
  </div>
</DocumentFragment>
`;
